openvdb/tree/Tree.h

```diff
@@ -1223,6 +1224,47 @@ class Tree: public TreeBase
     /// that this tree is about to be deleted.
     void releaseAllAccessors();
 
+    // NodeManager TBB body object used to deallocates tree nodes in parallel.
+    struct DeallocateTree
+    {
+        // clear the root node
+        void operator()(RootNodeType& root) const
+        {
+            root.clear();
+        }
+
+        // drop all children and set the child mask to off
+        template<typename NodeT>
+        void operator()(NodeT& node) const
+        {
+            auto* children = const_cast<typename NodeT::UnionType*>(node.getTable());
+            for (auto iter = node.beginChildOn(); iter; ++iter) {
+                delete children[iter.pos()].getChild();
+            }
+            const_cast<typename NodeT::NodeMaskType&>(node.getChildMask()).setOff();
+        }
+
+        // for allocated buffers, swap with a partially created (empty) buffer and de-allocate
+        template <typename ValueT>
+        inline void deallocateLeafBuffer(typename LeafNodeType::Buffer& buffer) const
+        {
+            if (!buffer.isOutOfCore() && !buffer.empty()) {
+                ValueT value;
+                typename LeafNodeType::Buffer newBuffer(openvdb::PartialCreate(), value);
+                buffer.swap(newBuffer);
+            }
+        }
+
+        // bool leaf buffers have no value buffer to be deallocated
+        template <>
+        inline void deallocateLeafBuffer<bool>(typename LeafNodeType::Buffer&) const { }
+
+        void operator()(LeafNodeType& leaf) const
+        {
+            this->deallocateLeafBuffer<ValueType>(leaf.buffer());
```

Comment:
To fix the build failures, maybe use tag dispatch instead:
```
this->deallocateLeafBuffer(leaf.buffer(), ValueType());
```
where `deallocateLeafBuffer` is defined as
```
template<typename ValueT>
void deallocateLeafBuffer(typename LeafNodeType::Buffer& buffer, const ValueT&) const
{
    ...
}
```
and then just overloaded for `bool`:
```
void deallocateLeafBuffer(typename LeafNodeType::Buffer& buffer, bool) const {}
```

Reply:
Ah yeah, funnily enough, my compiler doesn't complain about this.

How would you feel about adding a LeafBuffer::clear() method that contains the following and could be called from the deconstructor:

```
if (this->isOutOfCore()) {
    this->detachFromFile();
} else {
    this->deallocate();
}
```
I find I use this swapping trick quite a lot for things like streaming where I know I only want to read from each leaf node once. It would be nice to officially support destroying a leaf buffer.

Reply:
That sounds reasonable.  Maybe it deserves a separate PR though?

Reply:
Yes, I think so. Only reason I bundled this change in here is because my value type threw in my unit test and deallocation was causing that to fail. The scope has clearly changed though, so I'll adjust my unit test.